component/remote/vault/auth.go

```diff
@@ -0,0 +1,339 @@
+package vault
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/grafana/agent/pkg/flow/rivertypes"
+	vault "github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/api/auth/approle"
+	"github.com/hashicorp/vault/api/auth/aws"
+	"github.com/hashicorp/vault/api/auth/azure"
+	"github.com/hashicorp/vault/api/auth/gcp"
+	"github.com/hashicorp/vault/api/auth/kubernetes"
+	"github.com/hashicorp/vault/api/auth/ldap"
+	"github.com/hashicorp/vault/api/auth/userpass"
+)
+
+// An authMethod can configure a Vault client to be authenticated using a
+// specific authentication method.
+//
+// The vaultAuthenticate method will be called each time a new token is needed
+// (e.g., if renewal failed). vaultAuthenticate method may return a nil secret
+// if the authentication method does not generate a secret.
+type authMethod interface {
+	vaultAuthenticate(context.Context, *vault.Client) (*vault.Secret, error)
+}
+
+// AuthArguments defines a single authentication type in a remote.vault
+// component instance. These are embedded as an enum field so only one may be
+// set per AuthArguments.
+type AuthArguments struct {
+	AuthToken      *AuthToken      `river:"token,block,optional"`
+	AuthAppRole    *AuthAppRole    `river:"approle,block,optional"`
+	AuthAWS        *AuthAWS        `river:"aws,block,optional"`
+	AuthAzure      *AuthAzure      `river:"azure,block,optional"`
+	AuthGCP        *AuthGCP        `river:"gcp,block,optional"`
+	AuthKubernetes *AuthKubernetes `river:"kubernetes,block,optional"`
+	AuthLDAP       *AuthLDAP       `river:"ldap,block,optional"`
+	AuthUserPass   *AuthUserPass   `river:"userpass,block,optional"`
+	AuthCustom     *AuthCustom     `river:"custom,block,optional"`
+}
+
+func (a *AuthArguments) authMethod() authMethod {
+	switch {
+	case a.AuthToken != nil:
+		return a.AuthToken
+	case a.AuthAppRole != nil:
+		return a.AuthAppRole
+	case a.AuthAWS != nil:
+		return a.AuthAWS
+	case a.AuthAzure != nil:
+		return a.AuthAzure
+	case a.AuthGCP != nil:
+		return a.AuthGCP
+	case a.AuthKubernetes != nil:
+		return a.AuthKubernetes
+	case a.AuthLDAP != nil:
+		return a.AuthLDAP
+	case a.AuthUserPass != nil:
+		return a.AuthUserPass
+	case a.AuthCustom != nil:
+		return a.AuthCustom
+	}
+
+	panic("remote.vault: unreachable")
```

Comment:
Can we return an error here instead of panic? This feels like it could panic during an update.